oppijanumerorekisteri-domain/src/main/java/fi/vm/sade/oppijanumerorekisteri/repositories/impl/HenkiloViiteRepositoryImpl.java

```diff
@@ -1,56 +1,75 @@
-                        "      " + where + "\n" +
-                        "ORDER BY \"masterOid\", \"henkiloOid\""));
-        query.setResultTransformer(new AliasToBeanResultTransformer(HenkiloViiteDto.class));
-        return (List<HenkiloViiteDto>) query.list();
-    }
-    
-    private Where buildSqlConditions(HenkiloCriteria criteria, String h) {
-        Where where = new Where();
-        if (criteria.getHenkiloOids() != null) {
-            if (criteria.getHenkiloOids().isEmpty()) {
-                where.conditions.add("FALSE"); // ok in native queries
-            } else {
-                where.conditions.add(h+".oidhenkilo in (:oids)");
-                where.parameterSetters.add(q -> q.setParameterList("oids", criteria.getHenkiloOids(), StringType.INSTANCE));
+        List<HenkiloViiteDto> result;
+        if(!CollectionUtils.isEmpty(criteria.getHenkiloOids())) {
+            // Find master oids
+            List<String> masters;
+            // For large queries. This allows postgres to benefit from indices.
+            if(criteria.getHenkiloOids().size() >= 80) {
+                List<String> queryOids = criteria.getHenkiloOids().stream().collect(Collectors.toList());
+                masters = em.unwrap(Session.class).createSQLQuery("SELECT hv.master_oid \n" +
+                        "FROM henkiloviite hv \n" +
+                        "  INNER JOIN (\n" +
+                        "    SELECT '" + queryOids.get(0) + "' as query_oid\n" +
+                        queryOids.subList(1, queryOids.size()).stream().map(s -> "    UNION ALL SELECT '" + s + "'\n").collect(Collectors.joining())  +
+                        "  ) as hv_tmp on hv.master_oid = hv_tmp.query_oid OR hv.slave_oid = hv_tmp.query_oid\n").list();
             }
+            else {
+                masters = jpa().select(henkiloViite.masterOid.as("masterOid"))
+                        .from(henkiloViite)
+                        .where(henkiloViite.masterOid.in(criteria.getHenkiloOids())
+                                .or(henkiloViite.slaveOid.in(criteria.getHenkiloOids())))
+                        .fetch();
+            }
+
+            // Find all slaves for the master oids
+            result = jpa().select(Projections.bean(HenkiloViiteDto.class,
+                    henkiloViite.masterOid.as("masterOid"),
+                    henkiloViite.slaveOid.as("henkiloOid")))
+                    .from(henkiloViite)
+                    .where(henkiloViite.masterOid.in(masters))
+                    .fetch();
+        }
+        else {
+            // Find all
+            result = jpa().select(Projections.bean(HenkiloViiteDto.class,
+                    henkiloViite.masterOid.as("masterOid"),
+                    henkiloViite.slaveOid.as("henkiloOid")))
+                    .from(henkiloViite)
+                    .fetch();
         }
-        return where;
+
+        // Make sure all henkilos exist
```

Comment:
Olisihan se puhtaampaa jos nuo olisivat vain viittauksia mutta sen jälkeen täytyisi kuitenkin hakea kummankin viitatun henkilön oid erikseen henkilö taulusta. Ehkä paras olisi jos henkilö taulussa olis jollain tapaa one-to-many viittaus masteriin.